data_pipeline/twitter_ads_api/twitter_ads_api_pipeline.py

```diff
@@ -24,35 +30,101 @@ def get_client_from_twitter_ads_api(
     )
 
 
+def get_provenance(
+    source_config: TwitterAdsApiSourceConfig,
+    params_dict: Optional[Mapping[str, str]] = None
+) -> dict:
+    provenance = {
+        'imported_timestamp': datetime.utcnow().isoformat(),
+        'request_resource': source_config.resource
+    }
+    if params_dict:
+        provenance['request_params'] = [
+            {'name': key, 'value': value} for key, value in params_dict.items()
+        ]
+    LOGGER.debug('provenance: %r', provenance)
+    return provenance
+
+
 def get_bq_compatible_json_response_from_resource_with_provenance(
-    source_config: TwitterAdsApiSourceConfig
+    source_config: TwitterAdsApiSourceConfig,
+    params_dict: Optional[Mapping[str, str]] = None
 ) -> Any:
+    LOGGER.info("Getting data for resource: %s", source_config.resource)
+    LOGGER.info('params_dict: %r', params_dict)
     req = Request(
         client=get_client_from_twitter_ads_api(source_config=source_config),
         method="GET",
-        resource=source_config.resource
+        resource=source_config.resource,
+        params=params_dict
     )
-    provenance = {
-        'imported_timestamp': datetime.utcnow().isoformat(),
-        'request_resource': source_config.resource
-    }
-
     response = req.perform()
+    return remove_key_with_null_value({
+        **response.body,
+        'provenance': get_provenance(source_config, params_dict)
+    })
+
+
+def get_param_dict_from_required_params_and_value_from_bq(
+    source_config: TwitterAdsApiSourceConfig,
+    value_from_bq: str,
+    start_time: str,
+    end_time: str
+):
     return {
-        **remove_key_with_null_value(response.body),
-        'provenance': provenance
+        source_config.required_params['paramNameForBqValue']: value_from_bq,
+        source_config.required_params['paramNameForStartTime']: start_time,
+        source_config.required_params['paramNameForEndTime']: end_time
     }
 
 
+def iter_bq_compatible_json_response_from_resource_with_provenance(
+    source_config: TwitterAdsApiSourceConfig
+) -> Any:
+    if source_config.required_params:
+        value_list_from_bq = fetch_single_column_value_list_for_bigquery_source_config(
```

Comment:
it might be worth moving the contents of this if statement to a separate function? (wouldn't necessarily have to affect the test, which would still test this function)

Reply:
I am not sure why we should do that? In if statement only thing we are doing in this is calling functions.

Reply:
Lets keep this in mind to discuss it on Monday.